functions/fold.php

```diff
@@ -15,21 +15,27 @@
  *
  * @param  string $monoidClassName
  * @param  array  $collection
- * @return Monoid
+ * @return mixed
  */
-function fold(string $monoidClassName, $collection = null): Monoid {
+function fold(string $monoidClassName, $collection = null) {
     if (!is_a($monoidClassName, Monoid::class, true)) {
         throw new \InvalidArgumentException(
             sprintf('Class %s does not implement Monoid', $monoidClassName)
         );
     }
-    return reduce(
-        function (Monoid $acc, $curr) use ($monoidClassName) {
-            return concat($acc, new $monoidClassName($curr));
-        },
-        call_user_func("{$monoidClassName}::empty"),
-        $collection
-    );
+    $fold = function ($monoidClassName, $collection) {
+        return reduce(
+            function (Monoid $acc, $curr) use ($monoidClassName) {
```

Comment:
Maybe we can add the `: Monoid` type hint to this closure, what do you think?

Reply:
It can be added without problems, but in my editor it doesn't get syntax highlighting, so it looks weird, but it does add clarity.